tests/functionality/sopflow/selfcheck.cpp

```diff
@@ -112,23 +112,80 @@ struct SopflowFunctionalityTestParameters {
+  MPI_Comm comm;
+  int nprocs;
+  int logging_rank = 0;
+
+  SopflowFunctionalityTests(std::string testsuite_filename, MPI_Comm comm,
+                            int logging_verbosity = EXAGO_LOG_INFO)
+      : FunctionalityTestContext(testsuite_filename, logging_verbosity),
+        comm{comm} {
+    int my_rank;
+    auto rerr = MPI_Comm_rank(comm, &my_rank);
+    if (rerr)
+      throw ExaGOError("Error getting MPI rank number");
+
+    auto err = MPI_Comm_size(comm, &nprocs);
+    if (err) {
+      if (my_rank == logging_rank)
+        throw ExaGOError("Error getting MPI num ranks");
+      exit(0);
+    }
+  }
+
   void
   ensure_options_are_consistent(toml::value testcase,
                                 toml::value presets = toml::value{}) override {
+    int my_rank;
+    auto err = MPI_Comm_rank(comm, &my_rank);
+    if (err)
+      throw ExaGOError("Error getting MPI rank number");
+
+#if 0
+    int n_preset_procs;
+    set_if_found(n_preset_procs, presets, "n_procs");
+    int n_testcase_procs = -1;
+    set_if_found(n_testcase_procs, testcase, "n_procs");
+
+    if (-1 != n_testcase_procs) {
+      if (my_rank == logging_rank) {
+        std::stringstream errs;
+        errs << "Number of processes should be declared globally in the preset "
+                "area of the test suite TOML file, not inside each testcase.\n"
+             << "Testcase: " << testcase << "\nWith presets:\n"
+             << presets;
+        throw ExaGOError(errs.str().c_str());
+      }
+      exit(0);
+    } else if (nprocs != n_preset_procs) {
+      if (my_rank == logging_rank) {
+        std::stringstream errs;
+        errs << "SOPFLOW Functionality test suite found " << n_preset_procs
+             << " processes specified in the presets of the test suite TOML "
+                "file, but this test is being run with "
+             << nprocs << " processes.\nTestcase: " << testcase
+             << "\nWith presets:\n"
+             << presets;
+        throw ExaGOError(errs.str().c_str());
+      }
+      exit(0);
+      return;
+    }
+#endif
+
     auto ensure_option_available = [&](const std::string &opt) {
       bool is_available = testcase.contains(opt) || presets.contains(opt);
       if (!is_available) {
-        std::stringstream errs;
-        errs << "SOPFLOW Test suite expected option '" << opt
-             << "' to be available, but it was not found in this testsuite"
-             << " configuration:\n";
-        errs << testcase << "\nwith these presets:\n" << presets;
-        throw ExaGOError(errs.str().c_str());
+        if (my_rank == logging_rank) {
+          std::stringstream errs;
+          errs << "SOPFLOW Test suite expected option '" << opt
+               << "' to be available, but it was not found in this testsuite"
+               << " configuration:\n";
+          errs << testcase << "\nwith these presets:\n" << presets;
+          throw ExaGOError(errs.str().c_str());
```

Comment:
No exit(0) here in comparison...

Reply:
I put an exit in, but on looking at the code more carefully, I'm puzzled by the logic here. If you throw an exception, then you would expect that to get caught by the calling program so I'm not sure what exit(0) is supposed to do. I guess if it doesn't get caught then it should exit?

Reply:
I think if rank0 (or any rank) throws an exception, then all ranks will exit.

Exit(0) is likely harmless, but we have graceful cleanup in ExaGO utilities, so I think this is unnecessary

Reply:
I talked to Bill and got rid of the checks on rank 0. I'm adding a function that checks to see if the error is true anywhere and if it is, then everyone will throw an exception.

Reply:
I modified this slightly. I only do a global check on errors that are not MPI-related. Presumably if something like MPI_Comm_rank fails, then the Allreduce inside the check to see of there is a failure anywhere in the system is going to fail as well.